blosc/blosc.c

```diff
@@ -2117,6 +2118,34 @@ void blosc_set_splitmode(int mode)
   g_splitmode = mode;
 }
 
+/* Ensure global context safety post-fork via:
+ * 1. Acquire global context mutex pre-fork, blocks fork until global context
+ *    in released in other threads.
```

Comment:
Very good to add lots of comments to this complex problem!

Waiting for the lock pre-fork will cause deadlock in situations where a blosc worker thread wants to do a fork-exec. Maybe a bit exotic but a simple system("echo I was here >> /tmp/debug.log"); will cause deadlock. I am fine with that as long as we are aware of the consequences. It doesn't hurt to add error checking to the pthread_mutex_lock calls, since in some cases the deadlock can be discovered and EDEADLK be returned.

Reply:
This potential deadlock is resolved by the lazy-reinit strategy described in the alternate comment thread. Instead, it will become unsafe to `fork` while holding the `global_comp_mutex` and continue execution from the child process. As this mutex is an internal detail of `blosc.c`, and blosc isn't going to be using a fork-based pool strategy, this scenario shouldn't occur.